contracts/middleware/BaseMiddleware.sol

```diff
@@ -0,0 +1,39 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+import {Hooks} from "@uniswap/v4-core/src/libraries/Hooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {IHooks} from "@uniswap/v4-core/src/interfaces/IHooks.sol";
+import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {BeforeSwapDelta} from "@uniswap/v4-core/src/types/BeforeSwapDelta.sol";
+import {Proxy} from "@openzeppelin/contracts/proxy/Proxy.sol";
+
+contract BaseMiddleware is Proxy {
+    /// @notice The address of the pool manager
+    IPoolManager public immutable manager;
+    address public immutable implementation;
+
+    error FlagsMismatch();
+
+    constructor(IPoolManager _manager, address _impl) {
+        _ensureValidFlags(_impl);
+        manager = _manager;
+        implementation = _impl;
+    }
+
+    function _implementation() internal view override returns (address) {
+        return implementation;
+    }
+
+    // yo i wanna delete this function but how do i remove this warning
+    receive() external payable {
+        _delegate(_implementation());
+    }
+
+    function _ensureValidFlags(address _impl) internal view virtual {
```

Comment:
I wonder if the underlying implementation actually even needs to support the flags in this case given the middleware already does. What are your thoughts for requiring both addresses to be mined?

Reply:
initially, i did not require the implementation to be mined. however, when working on middleware-protect, i realized doing so would save a bit of gas on flag checks. as the dev would already have to mine a middleware, i don't think it's a big deal to enforce them to mine an implementation as well. also, it's likely that some implementations are already hooks and are already mined.

Reply:
there's a [caveat in middleware-protect](https://github.com/Uniswap/v4-periphery/pull/152/files#diff-871f5aec4b02f672d30cd275b29b6fae4a49586661e4ad6477547ae8d00c5435R91) that requires it to read the implementation permissions